### PR TITLE
Pre-load Annotation and FieldObject data before parsing it

### PR DESCRIPTION
**Pull request by Snuffleupagus**
By utilizing the `ObjectLoader` for this data we're able to invoke `AnnotationFactory.create` without having to handle the data-lookup potentially throwing `MissingDataException`s.
This could obviously lead to slightly more data being requested earlier than before[1], however it should be overall more efficient since we'll no longer risk needing to abort and re-parse annotations halfway through.

This patch also changes the `ObjectLoader`-class a little bit, to handle the /Annots-data, and it now supports the initial object being either a Dictionary or an Array.

---
[1] Note that `ObjectLoader.prototype.load` already has a fast-path, which avoids any lookup/parsing if the entire PDF document is already loaded.

### CODE REVIEW

**Comment by Snuffleupagus** [comment]
This appears to work fine, and I was even able to create a unit-test that fails if we don't pre-load the /Annots-data.

However, the `src/core/annotation.js` file is big and complex enough that I simply can *not* guarantee that there won't be a problem in some edge-case. If this does cause any regressions they'll be intermittent, and might be quite difficult to track down, since they'll be highly timing dependent[1]. On the other hand, the worst thing that could happen is that some annotations don't render but it shouldn't affect document rendering in general.

@calixteman How do you feel about this patch, could it be helpful despite the (small) risk outlined above?

---
[1] Which means that things such as the speed of the internet connection, the server, and/or the CPU itself will affect if an error occurs.

File: src/core/annotation.js [hunk 1]
```diff
@@ -74,40 +74,53 @@ import { OperatorList } from "./operator_list.js";
...
+      const [
         acroForm,
         xfaDatasets,
         structTreeRoot,
         baseUrl,
         attachments,
         globalColorSpaceCache,
-      ]) => ({
+      ] = await Promise.all([
+        pdfManager.ensureCatalog("acroForm"),
```
Comment (calixteman) [inline src/core/annotation.js:87]: nit: Maybe we could just create an array of promises without acroform, then await on the acroform one, add the objectLoader.load the array and then await on all promises.

**Review by calixteman — APPROVED** [review]
LGTM. Thank you.

File: src/core/annotation.js (hunk 1 above)
Comment (Snuffleupagus) [reply src/core/annotation.js]: Good idea.
I've attempted a slightly different approach, to keep everything in one place; please take a look.

**Comment by calixteman** [comment]
> This appears to work fine, and I was even able to create a unit-test that fails if we don't pre-load the /Annots-data.
> 
> However, the `src/core/annotation.js` file is big and complex enough that I simply can _not_ guarantee that there won't be a problem in some edge-case. If this does cause any regressions they'll be intermittent, and might be quite difficult to track down, since they'll be highly timing dependent[1]. On the other hand, the worst thing that could happen is that some annotations are skipped but it shouldn't affect document rendering in general.
> 
> @calixteman How do you feel about this patch, could it be helpful despite the (small) risk outlined above?
> 
> [1] Which means that things such as the speed of the internet connection, the server, and/or the CPU itself will affect if an error occurs.

Looking at the patch, I don't have the feeling that it's worst than before... I mean maybe we'll have new bugs because of that but maybe it's fixing some other bugs we never noticed because of their intermittency.
So let's take the risk and we'll see if we've some complains.

**Review by calixteman — APPROVED** [review]
LGTM. Thank you.

**Comment by Snuffleupagus** [comment]
It just occurred to me why we didn't already pre-load this data, and I really should have caught it a lot earlier; hence this patch unfortunately doesn't seem correct :-(

Please refer to the "NOTE" in the `ObjectLoader` JSDoc comment, which applies to the /Annots- and /Fields-entries since they will generally have /P-references that point to the /Page-dictionary and that in will turn contains /Parent-references that when walked will trigger the *entire* document to load.
https://github.com/mozilla/pdf.js/blob/6f052312d625224173db36d3e661657a89cf1865/src/core/object_loader.js#L45-L56
